### PR TITLE
Animaton: Drop into collapsed thought #2533

### LINKED ISSUE #2533: Animaton: Drop into collapsed thought
## Steps to Reproduce

```
- a
- b
- c
- d
- e
```

Drag and drop `a` into `d`.

## Current Behavior

`a` disappears and the rest of the thoughts translate into place:

https://github.com/user-attachments/assets/cce01bb1-0608-4f19-8722-c90b98c47a64

## Expected Behavior

`a` should translate to the destination thought and then disappear.

This should only occur if the destination is collapsed and thus the dragged thought is unmounted. If the destination thought is expanded (or will be expanded after the drop, such as when the cursor is on the dragged thought), then it should translate into place as normal (current behavior).

### PR DESCRIPTION
**Pull request by zukilover**
Resolves #2533

### CODE REVIEW

File: src/util/fauxAnimation.ts [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+    /** Remove the clone after the animation completes. */
+    const remove = () => {
+      clone.removeEventListener('animationend', remove)
+      if (clone.parentNode) clone.parentNode.removeChild(clone)
+    }
+    clone.addEventListener('animationend', remove)
+
+    // fallback cleanup in case animationend doesn't fire
+    window.setTimeout(remove, duration + 50)
```
Comment (raineorshine) [inline src/util/fauxAnimation.ts]: I think you can subscribe to `animationcancel` and avoid the fallback.

File: src/components/TreeNode.tsx [hunk 2]
```diff
@@ -264,6 +264,7 @@ const TreeNode = ({
                 : 'left {durations.layoutNodeAnimation} ease-out,top {durations.layoutNodeAnimation} ease-out',
         })}
         style={outerDivStyle}
+        data-thought-id={thoughtId}
```
Comment (raineorshine) [inline src/components/TreeNode.tsx:268]: Note that `thoughtId` is not unique for rendered thoughts: If the context view is open, the same thought may be rendered multiple times on screen at once. 

Instead, you can serialize `Path`, which is always unique.

File: src/util/fauxAnimation.ts [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+  })
+}
+
+export default startDropFauxAnimation
```
Comment (raineorshine) [inline src/util/fauxAnimation.ts]: On this project we follow a convention where the filename matches the default export name. This makes it easier to know where every function lives.

File: src/util/fauxAnimation.ts [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+/** Options for the faux drop animation. */
+type DropFauxOptions = {
+  /** The thought id of the dragged item to clone. */
+  fromThoughtId: ThoughtId
+  /** Destination rectangle (in viewport coordinates) to animate to. */
+  toRect: DOMRect
+  /** Duration in ms. */
+  duration?: number
+  /** Optional z-index for the faux clone. */
+  zIndex?: number
```
Comment (raineorshine) [inline src/util/fauxAnimation.ts]: Let's avoid adding optional parameters that are never used. We can always parameterize them in the future if needed, but better to avoid the additional complexity and creating the wrong expectations for the current code. When I saw these parameters, I assumed different values were being passed in different invocations, which turned out to be untrue. The existence of code implies its usage.

File: src/util/fauxAnimation.ts [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+/**
+ * Clones the dragged thought DOM node and animates it to the destination rect.
+ * Used when dropping into a collapsed destination where the original node unmounts immediately.
+ *
+ * Best-effort; silently aborts if DOM nodes cannot be found.
+ */
+const startDropFauxAnimation = ({ fromThoughtId, toRect, duration = 200, zIndex = 3 }: DropFauxOptions) => {
```
Comment (raineorshine) [inline src/util/fauxAnimation.ts]: Let's use the `durations` config and global z-index definition in the panda config rather than hardcoding values.

File: src/hooks/useDragAndDropSubThought.ts [hunk 6]
```diff
@@ -175,6 +176,36 @@ const drop = (props: DroppableSubthoughts, monitor: DropTargetMonitor) => {
...
 
+  // If the destination is collapsed, animate a faux clone of the dragged thought to the destination thought's position.
+  // Capture DOM positions before dispatching the move to avoid layout shifts during animation.
+  if (draggedItems.length === 1) {
+    const isDestinationExpanded = isPathExpanded(state, props.path)
+    if (!isDestinationExpanded) {
+      const destinationThoughtId = head(props.path)
+      if (destinationThoughtId) {
+        // select by aria-label, then filter by data attribute to avoid selector injection and errors
+        const nodes = document.querySelectorAll('[aria-label="tree-node"]') as NodeListOf<HTMLElement>
```
Comment (raineorshine) [inline src/hooks/useDragAndDropSubThought.ts]: I'm not sure what "selector injection and errors" means. Why can't a normal selector be used to select the thought?

File: src/hooks/useDragAndDropSubThought.ts [hunk 7]
```diff
@@ -175,6 +176,36 @@ const drop = (props: DroppableSubthoughts, monitor: DropTargetMonitor) => {
...
+        }
+
+        if (destinationEl) {
+          const toRect = destinationEl.getBoundingClientRect()
+          const fromThoughtId = head(draggedItems[0].path)
+          if (fromThoughtId) {
+            // kick off faux animation before DOM updates
+            fauxAnimation({ fromThoughtId, toRect })
+          }
+        }
```
Comment (raineorshine) [inline src/hooks/useDragAndDropSubThought.ts]: I would probably move all this logic to `fauxAnimation`. Notice that `useDragAndDropSubThought` does not currently have any DOM access. Putting all the direct DOM access in `fauxAnimation` keeps it more encapsulated.

File: src/util/fauxAnimation.ts (hunk 5 above)
Comment (raineorshine) [inline src/util/fauxAnimation.ts]: How about we call this `animateDroppedThought` or `animateToCollapsedParent`?

File: src/util/fauxAnimation.ts [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+  clone.style.left = `${fromRect.left}px`
+  clone.style.top = `${fromRect.top}px`
+  clone.style.width = `${fromRect.width}px`
+  clone.style.height = `${fromRect.height}px`
+  clone.style.zIndex = String(zIndex)
+  // drive translate distances with CSS variables to use keyframes
+  const dx = toRect.left - fromRect.left
+  const dy = toRect.top - fromRect.top
+  clone.style.setProperty('--faux-dx', `${dx}px`)
+  clone.style.setProperty('--faux-dy', `${dy}px`)
```
Comment (raineorshine) [inline src/util/fauxAnimation.ts]: Let's call this `--clone-dx` and avoid the "faux" terminology (here and elsewhere). "clone" is a bit more specific.

File: src/util/fauxAnimation.ts [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+  clone.style.position = 'fixed'
+  clone.style.pointerEvents = 'none'
+  clone.style.margin = '0'
+  clone.style.left = `${fromRect.left}px`
+  clone.style.top = `${fromRect.top}px`
+  clone.style.width = `${fromRect.width}px`
+  clone.style.height = `${fromRect.height}px`
+  clone.style.zIndex = String(zIndex)
+  // drive translate distances with CSS variables to use keyframes
+  const dx = toRect.left - fromRect.left
```
Comment (raineorshine) [inline src/util/fauxAnimation.ts]: Let's add `1em` to `dx` so that it looks like the thought is animating into a subthought position.

File: src/util/fauxAnimation.ts (hunk 1 above)
Comment (zukilover) [reply src/util/fauxAnimation.ts]: This has been updated.

File: src/components/TreeNode.tsx (hunk 2 above)
Comment (zukilover) [reply src/components/TreeNode.tsx:268]: This now uses `hashPath` instead of `thoughtId`

File: src/util/fauxAnimation.ts (hunk 3 above)
Comment (zukilover) [reply src/util/fauxAnimation.ts]: I've renamed to `animateDroppedThought`

File: src/util/fauxAnimation.ts (hunk 4 above)
Comment (zukilover) [reply src/util/fauxAnimation.ts]: It now accepts 2 params only:

```js
{
  /** Hashed path of the dragged item to clone (from hashPath). */
  fromPath: string
  /** Hashed path of the destination to animate to and track during animation (for handling layout shifts). */
  toPath: string
}
```

File: src/util/fauxAnimation.ts (hunk 5 above)
Comment (zukilover) [reply src/util/fauxAnimation.ts]: Updated to use panda config:

```js
const duration = durations.get('fast')
const zIndex = token('zIndex.cloneDroppedThought')
```

File: src/hooks/useDragAndDropSubThought.ts (hunk 6 above)
Comment (zukilover) [reply src/hooks/useDragAndDropSubThought.ts]: It was an overly cautious approach really. I've updated it so that it uses simpler selector:

```js
document.querySelector<HTMLElement>(`[data-path="${toPath}"]`)
```

File: src/hooks/useDragAndDropSubThought.ts (hunk 7 above)
Comment (zukilover) [reply src/hooks/useDragAndDropSubThought.ts]: I've moved it to `animateDroppedThought` util.

File: src/util/fauxAnimation.ts (hunk 5 above)
Comment (zukilover) [reply src/util/fauxAnimation.ts]: Renamed to `animateDroppedThought`

File: src/util/fauxAnimation.ts (hunk 8 above)
Comment (zukilover) [reply src/util/fauxAnimation.ts]: Renamed all faux terms to clone.

File: src/util/fauxAnimation.ts (hunk 9 above)
Comment (zukilover) [reply src/util/fauxAnimation.ts]: I added indent offset by using the `fontSize`. I see this is how indentation is calculated in `usePositionedThoughts`. Let me know if this looks OK.

File: src/components/TreeNode.tsx (hunk 2 above)
Comment (raineorshine) [reply src/components/TreeNode.tsx:268]: Thanks.

`data-thought-id` can be removed if it's no longer used.

File: src/util/animateDroppedThought.ts [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+
+    // read destination position from CSS properties
+    const destLeft = parseFloat(destinationEl.style.left) || 0
+    const destTop = parseFloat(destinationEl.style.top) || 0
+
+    // calculate translation distances
+    // add 1em indent to make it clear the thought is becoming a subthought
+    const computedStyle = window.getComputedStyle(source)
+    const fontSize = parseFloat(computedStyle.fontSize)
+    const indentOffset = fontSize // 1em
```
Comment (raineorshine) [inline src/util/animateDroppedThought.ts]: Rather than use `getComputedStyle`, you can do the calculation in CSS: `calc(${dx}px + 1em)`.

Or add a `1em` margin-left or translate if it doesn't conflict with existing style property.

`getComputedStyle` [forces reflow](https://gist.github.com/paulirish/5d52fb081b3570c81e3a#calling-getcomputedstyle) which may impact performance.

**Review by raineorshine — APPROVED** [review]
Looks good!

File: src/util/animateDroppedThought.ts (hunk 10 above)
Comment (zukilover) [reply src/util/animateDroppedThought.ts]: I've updated so that it uses `calc` for the indentation.

**Review by JoeDuong-dev — APPROVED** [review]
I’ve been testing this ticket and was able to reproduce the original issue before re-testing it on this branch. The drag-and-drop behavior is now working as expected. When the cursor is not on the dragged thought, the dragged thought correctly disappears into the parent and becomes its subthought. The user can expand the parent thought to view it, and the animation works as intended.

I also tested various edge cases with different text types, cursor positions, and drag-and-drop directions. The behavior remained consistent across all scenarios.

**Comment by karunkop** [comment]
> @karunkop Can you investigate the failed test run here: https://github.com/cybersemics/em/actions/runs/18637290170/attempts/1?pr=3264
> 
> Thanks!

Hi @raineorshine, I was not able to replicate this issue in the CI. [Here](https://github.com/karunkop/em/actions?query=branch%3Atest-flakiness-faux-animation) is the result for the checks that I did.
